pkg/controller/masteruserrecord/masteruserrecord_controller.go

```diff
@@ -116,6 +116,25 @@ func (r *ReconcileMasterUserRecord) Reconcile(request reconcile.Request) (reconc
 			logger.Error(err, "unable to add finalizer to MasterUserRecord")
 			return reconcile.Result{}, err
 		}
+		// migration for CRT-1075: add an annotation with the email address (same as for associated UserSignup resource)
+		if mur.Annotations == nil || mur.Annotations[toolchainv1alpha1.MasterUserRecordEmailAnnotationKey] == "" {
+			usersignup := &toolchainv1alpha1.UserSignup{}
+			if r.client.Get(context.TODO(), types.NamespacedName{
+				Namespace: mur.Namespace,
+				Name:      mur.Spec.UserID,
+			}, usersignup); err != nil {
+				logger.Error(err, "unable to get associated UserSignup")
+				return reconcile.Result{}, err
+			}
+			if mur.Annotations == nil {
+				mur.Annotations = map[string]string{}
+			}
+			mur.Annotations[toolchainv1alpha1.MasterUserRecordEmailAnnotationKey] = usersignup.Annotations[toolchainv1alpha1.UserSignupUserEmailAnnotationKey]
+			if r.client.Update(context.TODO(), mur); err != nil {
+				logger.Error(err, "unable to update the MUR with the email address annotation")
+				return reconcile.Result{}, err
+			}
+		}
```

Comment:
This migration shouldn't be in `MUR` controller, but in `UserSignup` controller - that controller is responsible for setting the proper annotation in MUR. We have this function: https://github.com/codeready-toolchain/host-operator/blob/master/pkg/controller/usersignup/masteruserrecord.go#L10-L41 for the purpose of migrating MUR - you can move the code there

Reply:
IMO it's a bit counter-intuitive to migrate the MasterUserRecords while processing the UserSignups, but since there's already a function that deals with that, then I will move the block above there 🤷‍♂️

Reply:
Not really. MUR is a secondary resource of the UserSignup controller. It means UserSignup controller actually manages MURs (creates, updates, deletes them). And MUR controller only watches MURs changes to manage its own secondary resources: UserAccounts. So it makes sense to migrate MURs in UserSignup controller.
But in practice I think it's fine to migrate resources in the place where it's easier to do.

Reply:
If I'm not mistaken, we do process every MUR in UserSignup controller as well.
The controller that manages secondary resources should ensure that the meta & spec information in the resources are up-to-date and matches the desired state - that's their purpose. 
In addition, to finish this particular the migration we need both resources: MUR and UserSignup - this is something we already have in UserSignup controller. Getting the UserSignup resource which is an owner of the MUR in MUR controller to make the MUR up-to-date looks like anti-pattern to me.

The only exception when we need to do the migration in the controller that watches the resources as the primary resource is UserSignup - the reason is that there is no controller that would manage UsesrSignups as secondary resources, so we need to do it in its controller.

Reply:
well, I missed your last comments while I was refactoring the code (see 9145b69), so please let me know if it's ok for you now. Basically, I added a block in the `migrateOrFixMurIfNecessary` to set the email address on the MUR if it's missing. And for sure, that having all the migration logic in one place is better. It's just that I missed that in my first PR 😬

Reply:
yes, it's perfect :-)  - thank you :+1: